bsd grep: panic when trying to split a matched string

for a specific string
`./pack/plugins/start/vim-go/ftplugin/gomod.vim:13:      \ | exe 'au! vim-go-gomod-buffer * <buffer>'"`
the splitting here doesn't work
https://github.com/vrothberg/vgrep/blob/9cc36833073ac6cae33dabdf87c92fcc3910cd90/vgrep.go#L251-L257
it fails and causes an index out of range error at line 256, because it didn't split into 3 parts
but only 1, lenght of spl is just 1
I added some print statements and this is the output
```bash
╰─➤  go run ~/go/src/github.com/vrothberg/vgrep/vgrep.go vim-go                                                                                    1 ↵
----------------Debugging------------------
>>> before split:
./pack/plugins/start/vim-go/ftplugin/gomod.vim:13:      \ | exe 'au! vim-go-gomod-buffer * <buffer>'"
>>> after splitting:
0 | ./pack/plugins/start/vim-go/ftplugin/gomod.vim:13:      \ | exe 'au! vim-go-gomod-buffer * <buffer>'"
----------------Debugging------------------
panic: runtime error: index out of range [1] with length 1

goroutine 1 [running]:
main.(*vgrep).splitMatch(0xc00005a240, 0xc000126000, 0x76, 0xc000140000, 0x11bce25, 0x32, 0xc000142000, 0x2e5, 0x2e6, 0x0)
        /Users/sjangra/go/src/github.com/vrothberg/vgrep/vgrep.go:264 +0x7ca
main.(*vgrep).grep(0xc00005a240, 0xc00004e540, 0x1, 0x1)
        /Users/sjangra/go/src/github.com/vrothberg/vgrep/vgrep.go:233 +0x308
main.main()
        /Users/sjangra/go/src/github.com/vrothberg/vgrep/vgrep.go:105 +0x46e
exit status 2
```
When it works successfully, debug output looks like this, for comparison
```
----------------Debugging------------------
>>> before split:
lruCache.go145func (cache *LRUCache) write(key string, data interface{}) (qdata, error) {
>>> after splitting:
0 | lruCache.go
1 | 145
2 | func (cache *LRUCache) write(key string, data interface{}) (qdata, error) {
----------------Debugging------------------
Index File        Line Content
    0 lru.go         9 cache := new(LRUCache)
    1 lruCache.go   23 type LRUCache struct {
    2 lruCache.go  107 func (cache *LRUCache) init(capacity int) {
    3 lruCache.go  112 func (cache *LRUCache) read(key string) (interface{}, error) {
    4 lruCache.go  125 func (cache *LRUCache) touch(key string) error {
    5 lruCache.go  145 func (cache *LRUCache) write(key string, data interface{}) (qdata, error) {
```

--- a/vgrep.go
+++ b/vgrep.go
@@ -241,7 +241,7 @@ func (v *vgrep) grep(args []string) {
 	}
 	v.matches = make([][]string, len(output))
 	for i, m := range output {
-		file, line, content := v.splitMatch(m, usegit, useripgrep, greptype)
+		file, line, content := v.splitMatch(m, greptype)
 		v.matches[i] = make([]string, 4)
 		v.matches[i][0] = strconv.Itoa(i)
 		v.matches[i][1] = file
@@ -254,8 +254,8 @@ func (v *vgrep) grep(args []string) {
 
 // splitMatch splits match into its file, line and content.  The format of
 // match varies depending if it has been produced by grep or git-grep.
-func (v *vgrep) splitMatch(match string, gitgrep bool, ripgrep bool, greptype string) (file, line, content string) {
-	if ripgrep {
+func (v *vgrep) splitMatch(match string, greptype string) (file, line, content string) {
+	if greptype == "RIP" {
 		// remove default color ansi escape codes from ripgrep's output
 		match = strings.Replace(match, "\x1b[0m", "", 4)
 	}

--- a/vgrep.go
+++ b/vgrep.go
@@ -191,7 +191,7 @@ func (v *vgrep) getGrepType() (grepType string) {
 	out, _ := v.runCommand([]string{"grep", "--version"}, "")
 	versionString := out[0]
 	// versionString = "grep (BSD grep) 2.5.1-FreeBSD"
-	versionRegex := regexp.MustCompile("\\(([[:alpha:]]+) grep\\)")
+	versionRegex := regexp.MustCompile(`\(([[:alpha:]]+) grep\)`)
 	// versionRegex matches to ["(BSD grep)", "BSD"], return "BSD"
 	grepType = versionRegex.FindStringSubmatch(versionString)[1]
 	return

--- a/vgrep.go
+++ b/vgrep.go
@@ -48,6 +48,14 @@ type vgrep struct {
 	waiter  sync.WaitGroup
 }
 
+// the type of underlying grep program
+const (
+	BSDGrep = "BSD"
+	GNUGrep = "GNU"
+	GITGrep = "GIT"
+	RIPGrep = "RIP"
+)
+
 // set in the Makefile
 var version string
 
@@ -218,7 +226,7 @@ func (v *vgrep) grep(args []string) {
 		}
 		cmd = append(cmd, args...)
 		cmd = append(cmd, ".")
-		greptype = "RIP"
+		greptype = RIPGrep
 	} else if usegit {
 		env = "HOME="
 		cmd = []string{
@@ -226,7 +234,7 @@ func (v *vgrep) grep(args []string) {
 			"grep", "-z", "-In", "--color=always",
 		}
 		cmd = append(cmd, args...)
-		greptype = "GIT"
+		greptype = GITGrep
 	} else {
 		env = "GREP_COLORS='ms=01;31:mc=:sl=:cx=:fn=:ln=:se=:bn='"
 		cmd = []string{"grep", "-ZIn", "--color=always"}
@@ -255,22 +263,22 @@ func (v *vgrep) grep(args []string) {
 // splitMatch splits match into its file, line and content.  The format of
 // match varies depending if it has been produced by grep or git-grep.
 func (v *vgrep) splitMatch(match string, greptype string) (file, line, content string) {
-	if greptype == "RIP" {
+	if greptype == RIPGrep {
 		// remove default color ansi escape codes from ripgrep's output
 		match = strings.Replace(match, "\x1b[0m", "", 4)
 	}
 	var seperator []byte
 	switch greptype {
-	case "BSD":
+	case BSDGrep:
 		seperator = []byte(":")
-	case "GIT", "GNU", "RIP":
+	case GITGrep, GNUGrep, RIPGrep:
 		seperator = []byte{0}
 	}
 	spl := bytes.SplitN([]byte(match), seperator, 3)
 	switch greptype {
-	case "BSD", "GIT":
+	case BSDGrep, GITGrep:
 		file, line, content = string(spl[0]), string(spl[1]), string(spl[2])
-	case "GNU", "RIP":
+	case GNUGrep, RIPGrep:
 		splline := bytes.SplitN(spl[1], []byte(":"), 2)
 		file, line, content = string(spl[0]), string(splline[0]), string(splline[1])
 	}

--- a/vgrep.go
+++ b/vgrep.go
@@ -187,6 +187,16 @@ func (v *vgrep) ripgrepInstalled() bool {
 	return installed
 }
 
+func (v *vgrep) getGrepType() (grepType string) {
+	out, _ := v.runCommand([]string{"grep", "--version"}, "")
+	versionString := out[0]
+	// versionString = "grep (BSD grep) 2.5.1-FreeBSD"
+	versionRegex := regexp.MustCompile("\\(([[:alpha:]]+) grep\\)")
+	// versionRegex matches to ["(BSD grep)", "BSD"], return "BSD"
+	grepType = versionRegex.FindStringSubmatch(versionString)[1]
+	return
+}
+
 // isVscode checks if the terminal is running inside of vscode.
 func isVscode() bool {
 	return os.Getenv("TERM_PROGRAM") == "vscode"
@@ -197,10 +207,10 @@ func (v *vgrep) grep(args []string) {
 	var cmd []string
 	var usegit bool
 	var env string
+	var greptype string // can have values , GIT, RIP, GNU, BSD
 
 	useripgrep := v.ripgrepInstalled() && !v.NoRipgrep
 	usegit = v.insideGitTree() && !v.NoGit
-
 	if useripgrep {
 		cmd = []string{
 			"rg", "-0", "--colors=path:none", "--colors=line:none",
@@ -208,6 +218,7 @@ func (v *vgrep) grep(args []string) {
 		}
 		cmd = append(cmd, args...)
 		cmd = append(cmd, ".")
+		greptype = "RIP"
 	} else if usegit {
 		env = "HOME="
 		cmd = []string{
@@ -215,22 +226,22 @@ func (v *vgrep) grep(args []string) {
 			"grep", "-z", "-In", "--color=always",
 		}
 		cmd = append(cmd, args...)
+		greptype = "GIT"
 	} else {
 		env = "GREP_COLORS='ms=01;31:mc=:sl=:cx=:fn=:ln=:se=:bn='"
 		cmd = []string{"grep", "-ZIn", "--color=always"}
 		cmd = append(cmd, args...)
 		cmd = append(cmd, "-r", ".")
+		greptype = v.getGrepType()
 	}
-
 	output, err := v.runCommand(cmd, env)
 	if err != nil {
 		fmt.Fprintf(os.Stderr, "searching symbols failed: %v\n", err)
 		os.Exit(1)
 	}
 	v.matches = make([][]string, len(output))
-
 	for i, m := range output {
-		file, line, content := v.splitMatch(m, usegit, useripgrep)
+		file, line, content := v.splitMatch(m, usegit, useripgrep, greptype)
 		v.matches[i] = make([]string, 4)
 		v.matches[i][0] = strconv.Itoa(i)
 		v.matches[i][1] = file
@@ -243,18 +254,28 @@ func (v *vgrep) grep(args []string) {
 
 // splitMatch splits match into its file, line and content.  The format of
 // match varies depending if it has been produced by grep or git-grep.
-func (v *vgrep) splitMatch(match string, gitgrep bool, ripgrep bool) (file, line, content string) {
+func (v *vgrep) splitMatch(match string, gitgrep bool, ripgrep bool, greptype string) (file, line, content string) {
 	if ripgrep {
 		// remove default color ansi escape codes from ripgrep's output
 		match = strings.Replace(match, "\x1b[0m", "", 4)
 	}
-	spl := bytes.SplitN([]byte(match), []byte{0}, 3)
-	if gitgrep && !ripgrep {
-		return string(spl[0]), string(spl[1]), string(spl[2])
+	var seperator []byte
+	switch greptype {
+	case "BSD":
+		seperator = []byte(":")
+	case "GIT", "GNU", "RIP":
+		seperator = []byte{0}
 	}
-	// the 2nd separator of grep is ":"
-	splline := bytes.SplitN(spl[1], []byte(":"), 2)
-	return string(spl[0]), string(splline[0]), string(splline[1])
+	spl := bytes.SplitN([]byte(match), seperator, 3)
+
+	switch greptype {
+	case "BSD", "GIT", "GNU":
+		file, line, content = string(spl[0]), string(spl[1]), string(spl[2])
+	case "RIP":
+		splline := bytes.SplitN(spl[1], []byte(":"), 2)
+		file, line, content = string(spl[0]), string(splline[0]), string(splline[1])
+	}
+	return
 }
 
 // getEditor returns the EDITOR environment variable (default="vim").

--- a/vgrep.go
+++ b/vgrep.go
@@ -267,11 +267,10 @@ func (v *vgrep) splitMatch(match string, gitgrep bool, ripgrep bool, greptype st
 		seperator = []byte{0}
 	}
 	spl := bytes.SplitN([]byte(match), seperator, 3)
-
 	switch greptype {
-	case "BSD", "GIT", "GNU":
+	case "BSD", "GIT":
 		file, line, content = string(spl[0]), string(spl[1]), string(spl[2])
-	case "RIP":
+	case "GNU", "RIP":
 		splline := bytes.SplitN(spl[1], []byte(":"), 2)
 		file, line, content = string(spl[0]), string(splline[0]), string(splline[1])
 	}

